### PR TITLE
[vcpkg] Added python script to generate all packages file list, added to azur…

### PR DESCRIPTION
**Pull request by yaoleo34**
A Python script is added in order to generate a full list of header files, called VCPKGHeaderDatabase.txt, this file is to be used for VS Code. A separate file is generated for all packages file list, called VCPKGDatabase.txt, this file is to be used for vcpkg website. This is achieved through Azure pipelines to generate these file lists at the end of the build.

### CODE REVIEW

**Comment by ras0219** [comment]
It looks like some leading slashes are being removed:

From the x86-windows VCPKGDatabase.txt:
```
xxhash:/debug/lib/xxhash.lib
xxhash:include/xxhash.h
xxhash:/lib/xxhash.lib
xxhash:/share/man/man1/xxhsum.1
xxhash:/share/xxhash/copyright
xxhash:/share/xxhash/vcpkg_abi_info.txt
zlib:/bin/zlib.pdb
zlib:/bin/zlib1.dll
zlib:/debug/bin/zlibd.pdb
zlib:/debug/bin/zlibd1.dll
zlib:/debug/lib/zlibd.lib
zlib:include/zconf.h
zlib:include/zlib.h
zlib:/lib/zlib.lib
zlib:/share/zlib/copyright
zlib:/share/zlib/usage
zlib:/share/zlib/vcpkg_abi_info.txt
zstd:/bin/zstd.dll
zstd:/bin/zstd.pdb
zstd:/debug/bin/zstdd.dll
zstd:/debug/bin/zstdd.pdb
zstd:/debug/lib/zstdd.lib
zstd:include/cover.h
zstd:include/zbuff.h
zstd:include/zdict.h
zstd:include/zstd.h
zstd:include/zstd_errors.h
zstd:/lib/zstd.lib
zstd:/share/zstd/COPYING
zstd:/share/zstd/LICENSE
zstd:/share/zstd/copyright
zstd:/share/zstd/vcpkg_abi_info.txt
zstd:/share/zstd/zstd-config-debug.cmake
zstd:/share/zstd/zstd-config-release.cmake
zstd:/share/zstd/zstd-config.cmake
```

**Comment by ras0219** [comment]
It would be best to include all files starting with `/include/`, instead of filtering by extension. This will catch headers like `#include <vector>` or `#include <boost/foo/magic.ipp>`

**Review by ras0219-msft — APPROVED** [review]
LGTM once the conditions are flipped :)

File: scripts/azure-pipelines/linux/azure-pipelines.yml [hunk 1]
```diff
@@ -52,3 +52,13 @@ jobs:
       PathtoPublish: '$(Build.ArtifactStagingDirectory)/failureLogs'
       ArtifactName: 'x64-linux port build failure logs'
     condition: failed()
+  - bash: |
+      python3 scripts/file_script.py /mnt/vcpkg-ci/installed/vcpkg/info/
+    displayName: 'Build a file list for all packages'
+    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
```
Comment (BillyONeal) [inline scripts/azure-pipelines/linux/azure-pipelines.yml:58]:
```suggestion
    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))

```

File: scripts/azure-pipelines/linux/azure-pipelines.yml [hunk 2]
```diff
@@ -52,3 +52,13 @@ jobs:
       PathtoPublish: '$(Build.ArtifactStagingDirectory)/failureLogs'
       ArtifactName: 'x64-linux port build failure logs'
     condition: failed()
+  - bash: |
+      python3 scripts/file_script.py /mnt/vcpkg-ci/installed/vcpkg/info/
+    displayName: 'Build a file list for all packages'
+    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
+  - task: PublishBuildArtifacts@1
+    displayName: 'Upload file lists for all packages'
+    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
```
Comment (BillyONeal) [inline scripts/azure-pipelines/linux/azure-pipelines.yml:61]:
```suggestion
    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))

```

File: scripts/azure-pipelines/osx/azure-pipelines.yml [hunk 3]
```diff
@@ -56,3 +56,13 @@ jobs:
       PathtoPublish: '$(Build.ArtifactStagingDirectory)/failureLogs'
       ArtifactName: 'x64-osx port build failure logs'
     condition: failed()
+  - bash: |
+      python3 scripts/file_script.py /Users/vagrant/Data/installed/vcpkg/info/
+    displayName: 'Build a file list for all packages'
+    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
```
Comment (BillyONeal) [inline scripts/azure-pipelines/osx/azure-pipelines.yml:62]:
```suggestion
    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))

```

File: scripts/azure-pipelines/osx/azure-pipelines.yml [hunk 4]
```diff
@@ -56,3 +56,13 @@ jobs:
       PathtoPublish: '$(Build.ArtifactStagingDirectory)/failureLogs'
       ArtifactName: 'x64-osx port build failure logs'
     condition: failed()
+  - bash: |
+      python3 scripts/file_script.py /Users/vagrant/Data/installed/vcpkg/info/
+    displayName: 'Build a file list for all packages'
+    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
+  - task: PublishBuildArtifacts@1
+    displayName: 'Upload file lists for all packages'
+    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
```
Comment (BillyONeal) [inline scripts/azure-pipelines/osx/azure-pipelines.yml:65]:
```suggestion
    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))

```

File: scripts/azure-pipelines/windows/azure-pipelines.yml [hunk 5]
```diff
@@ -64,3 +64,18 @@ jobs:
       PathtoPublish: '$(Build.ArtifactStagingDirectory)\failureLogs'
       ArtifactName: '${{ parameters.triplet }} port build failure logs'
     condition: failed()
+  - task: PowerShell@2
+    displayName: "Generating all packages files"
+    condition: and(succeeded(), eq(variables['Build.Reason'], 'PullRequest'))
```
Comment (BillyONeal) [inline scripts/azure-pipelines/windows/azure-pipelines.yml:69]:
```suggestion
    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))

```

**Review by BillyONeal — APPROVED** [review]

File: scripts/azure-pipelines/windows/azure-pipelines.yml [hunk 6]
```diff
@@ -64,3 +64,19 @@ jobs:
...
     condition: failed()
+  - task: PowerShell@2
+    displayName: "Generating all packages files"
+    condition: and(succeeded(), ne(variables['Build.Reason'], 'PullRequest'))
+
+    inputs:
+      targetType: inline
+      script: |
+        $env:VCPKG_DOWNLOADS = "D:\downloads"
+        ./vcpkg.exe fetch python3
```
Comment (BillyONeal) [inline scripts/azure-pipelines/windows/azure-pipelines.yml:75]: This is going to cause a merge conflict with a @ras0219 PR which removes this explicit variable set.

Also, why duplicate the fetch python3 call?

Comment (yaoleo34) [reply scripts/azure-pipelines/windows/azure-pipelines.yml:75]: The second fetch python3 call was to get the executable file path to execute python. 

Which variable set does the PR remove exactly?

**Comment by ras0219-msft** [comment]
LGTM, thanks Leo @yaoleo34 !

File: scripts/azure-pipelines/windows/azure-pipelines.yml (hunk 6 above)
Comment (BillyONeal) [reply scripts/azure-pipelines/windows/azure-pipelines.yml:75]: https://github.com/microsoft/vcpkg/pull/12198 <- This is the conflicting change. Your PR landed first so the comment goes to that one instead.

**Pull request merged**